services/system/StagedTx/service/src/lib.rs

```diff
@@ -0,0 +1,440 @@
+        pub id: u32,
+        pub account: AccountNumber,
+        pub accepted: bool,
+    }
+    impl Response {
+        #[primary_key]
+        fn by_id(&self) -> (u32, AccountNumber) {
+            (self.id, self.account)
+        }
+
+        #[secondary_key(1)]
+        fn by_responder(&self) -> AccountNumber {
+            self.account
+        }
+
+        fn upsert(id: u32, accepted: bool) {
+            let table = ResponseTable::new();
+            let response = table
+                .get_index_pk()
+                .get(&(id, get_sender()))
+                .map(|mut response| {
+                    response.accepted = accepted;
+                    response
+                })
+                .unwrap_or_else(|| Response {
+                    id,
+                    account: get_sender(),
+                    accepted,
+                });
+
+            table.put(&response).unwrap();
+        }
+    }
+
+    /// Initialize the staged-tx service
+    #[action]
+    fn init() {
+        let table = InitTable::new();
+        table.put(&InitRow {}).unwrap();
+
+        let updated = MethodNumber::from("updated");
+        Events::call().setSchema(create_schema::<Wrapper>());
+        Events::call().addIndex(DbId::HistoryEvent, SERVICE, updated, 0); // Index events related to specific txid
+        Events::call().addIndex(DbId::HistoryEvent, SERVICE, updated, 1); // Index events related to specific txid sender
+        Events::call().addIndex(DbId::HistoryEvent, SERVICE, updated, 2); // Index events related to specific proposer/accepter/rejecter
+    }
+
+    #[pre_action(exclude(init))]
+    fn check_init() {
+        let table: InitTable = InitTable::new();
+        check(
+            table.get_index_pk().get(&()).is_some(),
+            "service not initialized",
+        );
+    }
+
+    /// Proposes a new staged transaction containing the specified actions.
+    ///
+    /// * `actions` - The actions to be staged
+    #[action]
+    fn propose(actions: Vec<Action>) {
+        let new_tx = StagedTx::new(actions);
+
+        StagedTxTable::new().put(&new_tx).unwrap();
+
+        new_tx.emit(StagedTxEvent::PROPOSED);
+
+        // A proposal is also an implicit accept
+        accept(new_tx.id, new_tx.txid);
+    }
+
+    /// Indicates that the caller accepts the specified staged transaction
+    ///
+    /// Depending on the staging rules enforced by the auth service of the sender
+    /// of the staged transaction, this could result in the execution of the transaction.
+    ///
+    /// * `id`: The ID of the database record containing the staged transaction
+    /// * `txid`: The unique txid of the staged transaction
+    #[action]
+    fn accept(id: u32, txid: [u8; 32]) {
+        let staged_tx = StagedTx::get(id, txid);
+
+        Response::upsert(id, true);
+
+        staged_tx
+            .staged_tx_policy()
+            .accept(staged_tx.id, get_sender());
+
+        staged_tx.emit(StagedTxEvent::ACCEPTED);
```

Comment:
Is it okay for the `ACCEPTED` event to come after the `EXECUTED` and `DELETED` events?

Reply:
No, fixed.